drivers/flash/flash_rpi_pico.c

```diff
@@ -246,6 +275,29 @@ static int flash_rpi_write(const struct device *dev, off_t offset, const void *d
 	return 0;
 }
 
+static bool data_in_flash(const void *data)
```

Comment:
I am not sure this check is correct. SRAM is based at 0x20000000 and the XIP/SSI mapped region starts at 0x10000000, but it consists on many sub-regions that have different meanings.
Regions between 0x14000000-0x20000000 do not map to flash, and the 0x18000000-0x20000000 is actually range of SSI control registers.
The 0x20000000-0x13ffffff is actually 0x10000000 remapped with different attributes.
Btw. shouldn't this be rather named `data_in_xip_region`, it is not exactly in flash as the region contains possible cache check first.

Reply:
Should this check be used or is the always copy to ram solution by @Laczen preferred?

Reply:
The preference is on @yonsch , as far as I remember he is the platform maintainer.

I do not think that code size is an issue here, as the rpi uses external device for code, so having both functions is fine.
I guess that most of code would make ram-to-flash copy rather than flash-to-flash, which means that memcpy is just taking clocks; with a delay that the SPI processing would take it probably isn't that much on CPU. On the other side we do irq_lock here which means that we keep the irq_lock longer than probably needed in most cases.

I would stick with buffered/unbuffered separation.

I also wonder whether there should be added irq_unlock/lock within loop to allow irq processing between calls to write (@yonsch ?)

Reply:
I have missed the BSS usage by the flash_ram_buffer, which is PAGE_SIZE, it may be issue for some.
Anyway, in case of Nordic QSPI driver we have something like this:
https://github.com/zephyrproject-rtos/zephyr/blob/c33743ef955b066a5711097a6b0e184de1519b7f/drivers/flash/Kconfig.nordic_qspi_nor#L31-L41
where we also use intermediate buffer to copy data through and it is mostly enabled.